tests/tt_metal/tt_metal/perf_microbenchmark/9_dram_adjacent_read_remote_l1_write/test_dram_read_l1_write.cpp

```diff
@@ -0,0 +1,959 @@
+    for (auto core: all_cores) {
+        std::vector<uint32_t> result_vec;
+        tt_metal::detail::ReadFromDeviceL1(
+            device, core, cb_addr, num_tiles_cb * single_tile_size, result_vec);
+
+        uint32_t num_datum_per_block = block_h * block_w * num_datum_per_slice;
+        uint32_t tensor_slice_stride = core_id * num_datum_per_slice;
+        uint32_t last_block_offset = (num_blocks - 1) * num_datum_per_block * num_banks;
+        uint32_t start_index = tensor_slice_stride + last_block_offset;
+        uint32_t num_slices = block_h * block_w;
+
+        if (df == 0) {
+            auto result_bfp8 = unpack_bfp8_tiles_into_float_vec(result_vec, true, true);
+            auto input_bfp8 = unpack_bfp8_tiles_into_float_vec(input_vec, true, true);
+
+            for (uint32_t i=0; i < num_slices; ++i) {
+                uint32_t input_step = start_index + i * num_datum_per_slice * num_banks;
+                std::vector<float> input_slice(input_bfp8.begin() + input_step, input_bfp8.begin() + input_step + num_datum_per_slice);
+                uint32_t result_step = i * num_datum_per_slice;
+                std::vector<float> result_slice(result_bfp8.begin() + result_step, result_bfp8.begin() + result_step + num_datum_per_slice);
+
+                if (input_slice != result_slice) {
+                    return false;
+                }
+            }
+
+        } else {
+            auto result_bf16 = unpack_uint32_vec_into_bfloat16_vec(result_vec);
+            auto input_bf16 = unpack_uint32_vec_into_bfloat16_vec(input_vec);
+
+            for (uint32_t i=0; i < num_slices; ++i) {
+                uint32_t input_step = start_index + i * num_datum_per_slice * num_banks;
+                std::vector<bfloat16> input_slice(input_bf16.begin() + input_step, input_bf16.begin() + input_step + num_datum_per_slice);
+                uint32_t result_step = i * num_datum_per_slice;
+                std::vector<bfloat16> result_slice(result_bf16.begin() + result_step, result_bf16.begin() + result_step + num_datum_per_slice);
+
+                if (input_slice != result_slice) {
+                    return false;
+                }
+            }
+        }
+        core_id ++;
+    }
+    return true;
+}
+
+uint32_t get_dram_bandwidth(tt::ARCH arch) {
+    constexpr uint32_t GS_DRAM_BANDWIDTH_GB_PER_SEC = 100;
+    constexpr uint32_t WH_DRAM_BANDWIDTH_GB_PER_SEC = 384;
+
+    uint32_t dram_bandwidth_gb_per_sec = 0;
+    if (arch == tt::ARCH::WORMHOLE || arch == tt::ARCH::WORMHOLE_B0) {
+        dram_bandwidth_gb_per_sec = WH_DRAM_BANDWIDTH_GB_PER_SEC;
+    } else if (arch == tt::ARCH::GRAYSKULL) {
+        dram_bandwidth_gb_per_sec = GS_DRAM_BANDWIDTH_GB_PER_SEC;
+    }
+    return dram_bandwidth_gb_per_sec;
+}
+
+
+void get_dram_reader_core_coords_blackhole(
+    tt_metal::Device* device, CoreRangeSet& all_cores, std::vector<CoreCoord>& all_cores_ordered) {
+
+    // hardcoded for blackhole
+    uint32_t full_grid_size_x = 17;
```

Comment:
is this queryable through soc_descriptor?

can you add a comment on why this is BH specific?  I assume due to the physical layout?

Reply:
I don't think it queryable, haven't found an api to do that.
this is due to harvesting, I just need to get the max physical grid size, so that when moving the cores to the next cols, it won't fall off the grid.

Reply:
It should be queryable via `grid_size` from the soc descriptor:
```
const metal_SocDescriptor& soc_d = tt::Cluster::instance().get_soc_desc(device->id());
uint32_t full_grid_size_x = soc_d.grid_size.x
```

Reply:
changed with metal_SocDescriptor